### ISSUE TITLE
Attacking mnist using foolbox (FGSM)

### ISSUE THREAD

**Issue by SaiRaj07** [bethgelab/foolbox#175]
I am trying to implement gradient based attacks such as fgsm ,Bim,Jsma using foolbox. I took a look at the example in the documentation and accordingly implemented the attack on my own model. My model is defined in keras and accordingly I used foolbox, keras wrapper for the attack. So at a time I am able to generate adversary for only a single example, if I try multiple examples , I get an error saying, 
**Value Error : Cannot feed value of shape (1, 10000, 28, 28, 1) for Tensor 'conv2d_5_input:0', which has shape '(?, 28, 28, 1).**
Now I understand the error is with the input shape , but my variable explorer displays the shape as (10000,28,28,1) which is the correct shape and still I get the error.
I am attaching my code below:
```
import foolbox
import keras
import numpy as np
from keras import backend
from keras.models import load_model
from keras.datasets import mnist
from keras.utils import np_utils
from foolbox.attacks import SaliencyMapAttack
from foolbox.criteria import Misclassification
import matplotlib.pyplot as plt

########################################### Loading the model and preprocessing ###############################
backend.set_learning_phase(False)
model = keras.models.load_model('/home/labadmin/Mnist_Digits_Model_CNN.h5')
fmodel = foolbox.models.KerasModel(model, bounds=(0,1))
_,(images, labels) = mnist.load_data()
images = images.reshape(10000,28,28,1)
images= images.astype('float32')
images /= 255

######################################### Attacking the model ################################################
attack=foolbox.attacks.SaliencyMapAttack(fmodel, criterion=Misclassification())
adversarial=attack(images[12],labels[12]) # for single image
adversarial_all=attack(images,labels) # for all the images
adversarial =adversarial.reshape(1,28,28,1) #reshaping it for model prediction
model_predictions = model.predict(adversarial)
print(model_predictions)
######################################## Visualization #########################################################
images=images.reshape(10000,28,28)
adversarial =adversarial.reshape(28,28)

plt.figure()
plt.subplot(1,3,1)
plt.title('Original')
plt.imshow(images[12])
plt.axis('off')

plt.subplot(1, 3, 2)
plt.title('Adversarial')
plt.imshow(adversarial)
plt.axis('off')

plt.subplot(1, 3, 3)
plt.title('Difference')
difference = adversarial - images[124]
plt.imshow(difference / abs(difference).max() * 0.2 + 0.5)
plt.axis('off')
plt.show()
```

**Comment by jonasrauber** [comment]
The attacks expect a single input at at time. So if you want to create 10000 adversarials, you should loop over them yourself. Internally, the attacks might still send batches of images to the model, because an attack might need to evaluate the model many times to create a single adversarial.

**Comment by SaiRaj07** [comment]
As you said , I created a loop for 10000 images.  The file doesnt end running even after 1 hour and throws an error initially. It definitely should not take this long for fgsm attack which I currently tried. This error is thrown only for multiple examples situation

**UserWarning: GradientSignAttack did not find an adversarial, maybe the model or the criterion is not supported by this attack**
``
adversarial_images=[]
for i in range(len(images)) :
    attack=foolbox.attacks.FGSM(fmodel, criterion=Misclassification())
    img = images[i]
    label = labels[i]
    adversarial = attack(img, label)
    adversarial_images.append(adversarial)
`

**Comment by SaiRaj07** [comment]
As you said I created a loop for 10000 images and error shows up as follows :
**UserWarning: GradientSignAttack did not find an adversarial, maybe the model or the criterion is not supported by this attack**
This error shows up only for multiple examples  and file does not stop running even after 1 hour. It definitely should not take this long for fgsm attack.
```
adversarial_images=[]`
for i in range(len(images)) :
    attack=foolbox.attacks.FGSM(fmodel, criterion=Misclassification())
    img = images[i]
    label = labels[i]
    adversarial = attack(img, label)
    adversarial_images.append(adversarial)
```

**Comment by SaiRaj07** [comment]
Well on single images, it hardly takes 2 seconds and tried with few single images and could find adversarial for almost all. For all images its still running for around 30 minutes  since i started execution again. Currently after running for 30 minutes, there is only one warning.
Note:The same model takes 3 to 4 minutes to run for cleverhans (FGSM attack)

**Comment by jonasrauber** [comment]
2 seconds / image times 10000 images equals 5 hours ;-)

Cleverhans FGSM uses a fixed epsilon, i.e. fixed perturbation sizes. Foolbox (by default) tries to find minimal perturbations, so in the case of FGSM, it optimizes the epsilon for each image to find the closest adversarial.

**Comment by SaiRaj07** [comment]
1)Now its showing me another warning, User warning like the attack could not be implemented since the input is already adversarial and for attacks other than Fgsm, its throwing couldn't find an adversarial too many times as compared to fgsm. 

2)Also where i am saving the images, adversarial_images , it creates most images with type float32 but few are generated as Nonetype objects.

3) In case of all gradient based attacks, it will optimize the epilson for each image by trying to find the mnimum pertubations?? Can we modilfy and have an fix elipson like cleverhans because it takes a really long time for me to run .

**Comment by SaiRaj07** [comment]
1)Now its showing me another warning, User warning like the attack could not be implemented since the input is already adversarial and for attacks other than Fgsm, its throwing couldn't find an adversarial too many times as compared to fgsm.

2)Also where i am saving the images, adversarial_images , it creates most images with type float32 but few are generated as Nonetype objects.

3)In case of all gradient based attacks, it will optimize the epilson for each image by trying to find the mnimum pertubations?? Can we modilfy and have an fix elipson like cleverhans because it takes a really long time for me to run .

**Comment by maxdinech** [comment]
> User warning like the attack could not be implemented since the input is already adversarial

This is probably when the image already is a prediction error of the network (which is normal unless your network has an unlikely 0% error rate)

> few are generated as Nonetype objects.

The Nonetype objects come from failed attack: foolbox returns None rather than an image to make it very clear that the attack has failed.

**Comment by SaiRaj07** [comment]
Thank you. That helps alot to undertsand how foolbox works. When I am trying to use boundary attack , it throws an error as please install randomgen and even after install randomgen , it still throws the same error. Can you help me with that

**Comment by jonasrauber** [comment]
Maybe you installed it for the wrong python version or didn't restart your Jupyter notebook?

**Comment by Arkadeep-sophoIITG** [comment]
`adversarial_images=[]`
for i in range(len(images)) :
    attack=foolbox.attacks.FGSM(fmodel, criterion=Misclassification())
    img = images[i]
    label = labels[i]
    adversarial = attack(img, label)
    adversarial_images.append(adversarial)`

Why are you reinitializing the attack variable everytime in the for loop ?

**Comment by zhangdaoxian** [comment]
> It's not an error, it's a warning, informing you that for some images, the attack couldn't find an adversarial. This can sometimes happen because FGSM is a very basic attack.
> 
> How long does the attack take on your model for individual images?
> 
> If you get the warning for many images that might indicate that something with your gradients or model wrapping is wrong. For those images where the attack fails, the attack takes quite some time, because it tries many perturbation sizes before giving up.

I got the warning for most of my images, how can I know whats wrong with my gradients or model wrapping?
`from keras.models import load_model`
`model_path = r'.\Model_toplayer_softmax.h5'`
`ori_model = load_model(model_path)`
`    model=foolbox.models.KerasModel(ori_model,bounds=(0,1),predicts='probabilities')`
`    FGSM_attack = GradientSignAttack(model,criterion=Misclassification())`

**Comment by wielandbrendel** [comment]
That's hard to tell without knowing more details about your model. One alternative is to use decision-based attacks like the BoundaryAttack which does not rely on gradients.

**Comment by zhangdaoxian** [comment]
> That's hard to tell without knowing more details about your model. One alternative is to use decision-based attacks like the BoundaryAttack which does not rely on gradients.

Thanks a lot for your reply!
Considering your suggestion, I wonder if this is true that basic gradient-based attack like FGSM dosen't work for some models?
If this is helpful: I'm trying to attack a model I trained transfer learned from VGG-16, following [this Keras Example](https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html).
